src/nodes/NodeConnectionManager.ts

```diff
@@ -0,0 +1,716 @@
+          )}`,
+        );
+        // Creating the connection and set in map
+        return await this.establishNodeConnection(targetNodeId, lock);
+      });
+    }
+  }
+
+  /**
+   * Strictly a helper function for this.createConnection. Do not call this
+   * function anywhere else.
+   * To create a connection to a node, always use createConnection, or
+   * withConnection.
+   * This only adds the connection to the connection map if the connection was established.
+   * @param targetNodeId Id of node we are establishing connection to
+   * @param lock Lock associated with connection
+   * @returns ConnectionAndLock that was added to the connection map
+   */
+  protected async establishNodeConnection(
+    targetNodeId: NodeId,
+    lock: RWLock,
+  ): Promise<ConnectionAndLock> {
+    const targetAddress = await this.findNode(targetNodeId);
+    // If the stored host is not a valid host (IP address), then we assume it to
+    // be a hostname
+    const targetHostname = !networkUtils.isHost(targetAddress.host)
+      ? (targetAddress.host as string as Hostname)
+      : undefined;
+    const targetHost = await networkUtils.resolveHost(targetAddress.host);
+    const connection = await NodeConnection.createNodeConnection({
+      targetNodeId: targetNodeId,
+      targetHost: targetHost,
+      targetHostname: targetHostname,
+      targetPort: targetAddress.port,
+      forwardProxy: this.fwdProxy,
+      keyManager: this.keyManager,
+      nodeConnectionManagerInterface: this.getInterface(targetNodeId),
+      connConnectTime: this.connConnectTime,
+      logger: this.logger.getChild(`${targetHost}:${targetAddress.port}`),
+      clientFactory: async (args) =>
+        GRPCClientAgent.createGRPCClientAgent(args),
+    });
+    // Creating TTL timeout
+    const timer = setTimeout(async () => {
+      await this.destroyConnection(targetNodeId);
+    }, this.connTimeoutTime);
+    // Add it to the map of active connections
+    const connectionAndLock = { connection, lock, timer };
+    this.connections.set(targetNodeId.toString(), connectionAndLock);
+    return connectionAndLock;
+  }
+
+  /**
+   * Removes the connection from the connection man and destroys it.
+   * @param targetNodeId Id of node we are destroying connection to
+   */
+  protected async destroyConnection(targetNodeId: NodeId): Promise<void> {
+    /**
+     * Get connAndLock from connections
+     * if connAndLock != null:
+     *   if conn != null:
+     *     acquire lock
+     *     destroy the conn
+     *     // if we destroy here, we need to ensure that any action using the conn must also
+     *     // acquire the lock, otherwise we could be performing an action whilst the conn
+     *     // is being destroyed
+     *     // we also need to consider a thread that may be waiting for the lock to be released
+     *     // if we destroy the conn and the lock, then what happens to the waiting thread?
+     *     delete from connections
+     */
```

Comment:
We should be able to remove this block comment now. These were old notes by me.

Reply:
Removed